### ISSUE TITLE
Issue with directory layout of compiled Java classes and resources, 'run' task and Jigsaw

### ISSUE THREAD

**Issue by zyxist** [gradle/gradle#4733]
Hello,

I'm developing a custom Gradle plugin that provides a support for building Java 9 modular (Jigsaw)  applications, until Gradle gets a native support for them. I encountered an issue in the directory layout of compiled Java classes and processed resources, that leads to problems with `run` task.

The original issue was reported here: https://github.com/zyxist/chainsaw/issues/22 - I think you will fall into exactly the same problem, while implementing https://github.com/gradle/gradle/issues/3618

The problem is that `build/classes` and `build/resources` are distinct directories due to the fact that they are generated by two distinct Gradle tasks. `run` task starts the application by putting the two directories into the classpath, and it does not use a packaged JAR archive. I came to a conclusion that this was done to speed up the execution.

Now, if we want to use Jigsaw here, we must construct `--module-path`. The naive solution is to put both directories there, but it doesn't work. JVM would see a module descriptor in `build/classes`, and would not consider `build/resources` to be a part of that module. So, we need to use `--patch-modules` to add `build/resources` directory to our module.

Here comes the real issue. Suppose that you have a following directory layout:

```
/src/main/java/
   /com/example/MyClass.java
   /module-info.java
/src/main/resources/com/example/images
  /logo.jpg
```

The important thing is that `com.example.images` is present only in the project resources, and not in the Java sources. Now we want to open this package to the world in the module descriptor:

```java
module com.example {
   opens com.example.images;
}
```

It works, if we run the application manually from a JAR file. However, Gradle with Jigsaw flags added fails here.

### Expected Behavior

It is possible to open a package that is present only in the project resources. Gradle is able to build and run such a project.

### Current Behavior

Gradle execution fails:

```
java.lang.module.InvalidModuleDescriptorException: Package com.example.images not found in module
```

### Context

The main cause is that the module descriptor seems to be processed by JVM earlier than `--patch-module` flags, and this package is simply not yet visible.

Do you think this behavior would require deeper changes to the way how Gradle works? Or maybe the simplest solution would be to make `run` task dependent on a `jar` task, and run the application from the JAR archive?

**Comment by Ealrann** [comment]
Hello,

I encountered this issue in my project too. 
For the record, some other people talk about that [here](https://stackoverflow.com/questions/49678324/how-to-work-with-resources-in-java-9-modules?answertab=active#tab-top), and they provided a test project here:

https://github.com/michas2/java_9_resource_test

**Comment by rgoldberg** [comment]
The diagnosis of this problem is incorrect.

`/src/main/resources/com/example/images` is not a package named `com.example.images`.  It's just a directory in a resources directory.

Why are you trying to use `opens`?  Why do you `want to open this package to the world`? What problem do you think that will solve?

**Comment by Ealrann** [comment]
> Why are you trying to use `opens`? Why do you `want to open this package to the world`? What problem do you think that will solve?

I don't know about Zyxist, but for me, I'm using the ```opens``` to read my resources.

```module.getResourceAsStream(path)```
=> If the path of the resource is not ```opens``` in the _module-info.java_, the code above return null.

My understanding of all of that is limited, but it seems match to the [javadoc of this method](https://docs.oracle.com/javase/9/docs/api/java/lang/Module.html#getResourceAsStream-java.lang.String-): 

> [...] the resource can only be located by the caller of this method when the package is [open](https://docs.oracle.com/javase/9/docs/api/java/lang/Module.html#isOpen-java.lang.String-) to at least the caller's module.

**Comment by jjohannes** [comment]
Native support for Java modules is now available in the latest `6.4` nightly - see https://github.com/gradle/gradle/issues/890#issuecomment-603289940 for details. Please comment there for further feedback and questions.
